knitr/planetary_motion/planetary_motion.rmd

```diff
@@ -0,0 +1,735 @@
+$$
+  q_\mathrm{obs} \sim \mathrm{Normal}(q, \sigma I),
+$$
+with $\sigma$ the standard deviation and $I$ the $2 \times 2$  identity matrix.
+
+## Simulation
+
+The simulation can be done in Stan, using one of its numerical integrators.
+At $t = 0$, $q_0 = (1, 0)$ and $p_0 = (0, 1)$,
+and we set $k = 1$.
+```{r, fig.width=4, fig.height=3}
+mod <- cmdstan_model("model/planetary_motion_sim.stan")
+
+n <- 40
+sim <- mod$sample(data = list(n = n, sigma_x = 0.01, sigma_y = 0.01), 
+                  chains = 1, iter_warmup = 1, 
+                  iter_sampling = 2, seed = 123)
+
+r_sim <- read_stan_csv(sim$output_files())
+simulation <- rstan::extract(r_sim, pars = "q_obs")$q_obs[1, , ]
+
+q_x <- simulation[, 1]
+q_y <- simulation[, 2]
+
+q_obs <- array(NA, c(n, 2))
+q_obs[, 1] <- q_x
+q_obs[, 2] <- q_y
+
+sub_set <- seq(from = 1, to = 40, by = 2)
+plot <- ggplot(data = data.frame(q_x = q_x[sub_set], 
+                                 q_y = q_y[sub_set], 
+                                 time = sub_set), 
+               aes(x = q_x, y = q_y, label = time)) +  theme_bw() +
+  geom_text()
+plot
+```
+
+# Fitting a simple model and diagnosing inference
+
+A first attempt at fitting the complete model fails spectacularly: the chains do not converge and take a long time to run.
+This is an invitation to start with a simpler model.
+A useful simplification is more manageable but still exhibits the challenges we encounter with the complete model.
+The hope is that a fix in this simple context translates into a fix in a more sophisticated setting.
+This turns out to be quite true here where a simple one-parameter model allows us to understand the fundamentally multimodal nature of the model.
+
+There are many ways to simplify a model.
+A general approach is to fix some of the parameters, which we can easily do when working with simulated data.
+We fix all the parameters, except for $k$.
+We now want to characterize the posterior distribution,
+$$
+   p(k \mid q_\mathrm{obs}),
+$$
+by fitting the model with Stan.
+Since $k$ is a scalar, we could use quadrature but our goal is to understand the pathologies that frustrate our inference algorithm, so we stick to MCMC.
+As a prior, we use
+$$
+  k \sim \mathrm{Normal}^+(0, 1),
+$$
+which is fairly uninformative.
+
+We fit 8 chains in parallel.
+```{r }
+chains <- 8
+```
+
+```
+fit <- mod$sample(data = list(n = n, q_obs = q_obs),
+                  chains = chains, parallel_chains = chains,
+                  iter_warmup = 500,
+                  iter_sampling = 500,
+                  seed = 123, save_warmup = TRUE)
+```
+
+```{r }
+# The model takes a while to run, so we read in the saved output.
+r_fit1 <- readRDS("saved_fit/fit1.RDS")
+get_elapsed_time(r_fit1)
+```
+
+The first notable pathology is that some of the chains take a much longer time to run. The difference is not subtle...
+
+Let's examine the summary.
+```{r }
+summary(r_fit1, pars = c("lp__", "k"), probs = c())[1]
```

Comment:
Assuming this is going to cmdstanr, `fit$summary(pars = c("lp__", "k"))`

Reply:
Just one minor edit, you can use either of these: 
```r
fit$summary(c("lp__", "k")) 
fit$summary(variables = c("lp__", "k"))
```

We ended up going with `variables` instead of `pars` because that's what the posterior packages uses and also it makes more sense than `pars` (since there are generated quantities and lp__ too).

Reply:
Makes sense. Also, for the Ising and spin glass models, the variables I sample are technically not parameters, so the more general term is sensible.